### ISSUE TITLE
Re-architect widget-load and data transfer workflow

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/node-red-dashboard#1122]
### Description

Capturing this in a new issue after lengthy discussions and issues captured here:

- https://discourse.nodered.org/t/how-to-sync-the-full-config-between-server-and-client-side-state/89252/80
- https://github.com/FlowFuse/node-red-dashboard/issues/1117

#### Widget Load Loses State

The problem can be seen in the Events Architecture [here](https://dashboard.flowfuse.com/contributing/guides/events.html#loading-event-flow):

<img width="1546" alt="Screenshot 2024-07-24 at 14 22 57" src="https://github.com/user-attachments/assets/d2e3da2e-1d59-4c3b-8e99-9c62bf715625">

Note that in the `widget-load` event, there is no reference to the `state` store, so when a user navigates away from a page, updates to that store can occur, but the `widget-load` events is not aware of them.

Loosely thinking, we need to have something like:

<img width="1277" alt="Screenshot 2024-07-24 at 14 23 39" src="https://github.com/user-attachments/assets/772ab9b7-c409-49db-a0ff-fcce83e3d7a9">

#### Data Store Changes

There is also merit here in swapping how the `datastore` is used - rather than being a full `msg` store - we could treat it like `state` in that it's a singular, cumulative reference to the `data` for a single widget. For example, if `msg.payload` and `msg.otherProperty` are sent in _different_ messages, they would still be stored in the same place, and would be replayed accordingly.

**Comment by colinl** [comment]
I am confused about the relation between the widget-load event and mounted(). In a third party node in mounted() the state store *is* passed in, so it isn't necessary to include in widget-load.

**Comment by joepavitt** [comment]
`mounted()` is just a VueJS event to saw the HTML has loaded.

I think the `state` passed in that you're referring to is on the component definition, and I'll likely be removing it. It was part of an old design pattern that can be cleared out as only one widget (`ui-button`) still uses it, and it doesn't actually need to.

`widget-load` is the event that the component uses to tell the server-side _"I'm ready, send me the latest info you have please"_, which, up until this issue, has _only_ been the latest `msg` received by that widget.

What we're now proposing is introducing the server-side `state` here  (from the `statestore`) too so that the widget is aware of any changes made between the initial configuration (being sent in `ui-config`) and the `widget-load` event

**Comment by colinl** [comment]
> It was part of an old design pattern that can be cleared out as only one widget (`ui-button`) still uses it, and it doesn't actually need to.

Isn't it the mechanism by which, at the moment, third party nodes receive the node config?  Won't removing that break all existing third party nodes? Or is it just me doing it the wrong way?

**Comment by joepavitt** [comment]
Node config comes in via `props` on a component at the moment - this will still be the case

**Comment by colinl** [comment]
Oh, I see, I presume that is what this does
```
    props: {
        /* do not remove entries from this - Dashboard's Layout Manager's will pass this data to your component */
        id: { type: String, required: true },
        props: { type: Object, default: () => ({}) },
        state: { type: Object, default: () => ({ enabled: false, visible: false }) }
    },
```
So props and state are automagically available, but when navigating away and back again the state is not updated to the latest, hence the problem you are addressing here.  My nodes work because I am picking up the state in mounted(), which is re-run on navigation away and back again, giving me the latest values.  Are you sure it is not better just to make that the recommended way?

**Comment by joepavitt** [comment]
huh - I'd completely forgot I implemented that, maybe I need to re-consider this again... back to the drawing board

**Comment by arturv2000** [comment]
This behaviour was also mentioned in #871 

From my understanding, not all widgets generate a `widget-load` event, for example `ui-button`, this behaviour should be standardized for all widgets (at least core widgets), upon being mounted (loaded, either via page refresh or via page/tab change) the configuration `props` or `msg.ui_update` should be updated for each widget in the page, if no configuration exists in the store maybe a `empty` message should also be sent, to indicate no configuration pending for example...

Maybe a better place would be on file `ui_base.js` in the function `async function onLoad (conn, id, msg) {`, to add all the contents of the statestore to a `msg.ui_update` and emit a `widget-load` event that will be catch by `data-tracker.mjs` that will send to the widgets, for this to work the widgets must have the `onLoad` handler defined and handling the message, or `data-tracker` on seeing the message has `ui_update` calls `onDynamicProperties` event.

**Comment by colinl** [comment]
> if no configuration exists in the store

Which store?

I don't know what is wrong with the current implementation. `mounted` provides the latest props and state store, `widget-load` provides the data store, if any.

> not all widgets generate a `widget-load` event, for example `ui-button`
A button never has anything in the data store, so that is ok.

**Comment by joepavitt** [comment]
> this behaviour should be standardized for all widgets (at least core widgets)

This is my plan for this issue, and what I'm working on as we speak

> Maybe a better place would be on file ui_base.js in the function async function onLoad (conn, id, msg) {, to add all the contents of the statestore to a msg.ui_update and emit a widget-load event that will be catch by data-tracker.mjs that will send to the widgets, for this to work the widgets must have the onLoad handler defined and handling the message, or data-tracker on seeing the message has ui_update calls onDynamicProperties event.

Precisely what I'm doing 😄

**Comment by joepavitt** [comment]
> I don't know what is wrong with the current implementation

If you pass a `ui_update` to a node that's not on a page you're viewing now, and then navigate to that page, you will not see the effects of it. You will, if you then refresh the page. The `widget-load` event contains zero reference to anything in our server-side `statestore` (where we save dynamic options)

**Comment by colinl** [comment]
> If you pass a `ui_update` to a node that's not on a page you're viewing now, and then navigate to that page, you will not see the effects of it.

You do, provided the ui_update has updated the state store in the server.

**Comment by joepavitt** [comment]
I guarantee you do not @colinl - we've had many users report the bugs and I can see that problem locally. By "navigate to that page", I mean via the sidebar btw. If you navigate directly to it, it will work because the `ui-config` event is fired then.

**Comment by arturv2000** [comment]
> Which store?

Regarding `ui_update` would say is the `statestore` -> `require('../store/state.js')`

> You do, provided the ui_update has updated the state store in the server.

The `ui_update` is currently being updated on the `statestore` on server side when a new message is sent by node-red, but it is not currently being restored to re-send the configuration to the widget, that is the point of this and some other issues / discussions.

In most cases on a page refresh it works since the function `function emitConfig (socket) {` in `ui_base.js` handles that correctly (to get the values from `statestore`, to send the configuration to the widget but in the `props` property, if the widget is not expecting to have the configuration in question in the `props` (like in the ui-dropdown related to issue #1117) it never receives it even in a page refresh.

**Comment by joepavitt** [comment]
@arturv2000 FYI I've just settled on some large changes to how we're handling front-end `dynamic` properties, will be providing a new API to make this _a lot_ easier. I'm updating docs now, and will push a draft PR. 

I'll be sweeping through myself and updating any existing use of `dynamic`, but wanted to give you a heads up as I know you were working your way through a few of these.

**Comment by colinl** [comment]
> I guarantee you do not @colinl - we've had many users report the bugs

I was referring to my nodes, which do not suffer from the problem. Updates saved in the state store are seen in mounted() on a page change.

**Comment by joepavitt** [comment]
Okay, I don't know what you're doing different to core then in order to achieve that, but I have seen a flaw in the architecture based on the bug reports, and I'm fixing now.

If I recall correctly, you were using `datastore`, and not `statestore` - the former works, the latter does not

**Comment by colinl** [comment]
Aargh, now I am confused. I do see the problem on a new node I am working on, but not on the classic gauge node.  I will have to look and see what the difference is.

**Comment by joepavitt** [comment]
Don't worry, a fix for all of this is incoming

**Comment by colinl** [comment]
> If I recall correctly, you were using `datastore`, and not `statestore` - the former works, the latter does not

Yes you are right, that explains it.  Sorry again.

> Don't worry, a fix for all of this is incoming

If it is likely to break existing nodes it would be good to have a gap between merging into main and release so testing and updates can be done.

**Comment by joepavitt** [comment]
> If it is likely to break existing nodes it would be good to have a gap between merging into main and release so testing and updates can be done.

If all goes to plan, this _shouldn't_ break anything. It'll just provide a new API to interface with, and argument for `widget-load` which will help you load the correct data

**Comment by arturv2000** [comment]
> @arturv2000 FYI I've just settled on some large changes to how we're handling front-end `dynamic` properties, will be providing a new API to make this _a lot_ easier. I'm updating docs now, and will push a draft PR.
> 
> I'll be sweeping through myself and updating any existing use of `dynamic`, but wanted to give you a heads up as I know you were working your way through a few of these.

Yes, i have been trying, there is one PR #1059 related to that, that would presumably solve the issue for `ui-button`, for the rest of the widgets would need to be updated / Verified for similar changes. That PR was an attempt to solve the issue without too many changes but obviously not the ideal way to do that...

I will close that PR and wait for your implementation to do some testing/possible feedback when the PR starts coming.

**Comment by joepavitt** [comment]
>  That PR was an attempt to solve the issue without too many changes but obviously not the ideal way to do that...

yeah, I've just had to rip the band-aid with this one, really appreciate your time on this though

**Comment by colinl** [comment]
Once the work is complete, would it be possible for you to extend the example node so I can see the recommended way of handling it in third party nodes?  Possibly just a config field that is displayed in the widget and can be updated via ui_update.

**Comment by joepavitt** [comment]
> Would it be possible for you to extend the example node so I can see the recommended way of handling it in third party nodes?

Of course

**Comment by colinl** [comment]
Great, thanks.

**Comment by joepavitt** [comment]
I will add example into https://github.com/FlowFuse/node-red-dashboard-2-ui-example

But we'll also heavily benefit from https://github.com/FlowFuse/node-red-dashboard/pull/1099 here too

So, this PR, whilst ready, I won't merge/review yet, I'll hold out for next release

**Comment by arturv2000** [comment]
Will do some testing tonight or during the weekend to see if my main concerns are addressed.

Did a small test last night with a `ui-button` and the issue of `enabled` and `class` been correctly updated when changing to the widget visible tab seems OK.

**Comment by colinl** [comment]
I know how to build and run the dashboard locally, but I don't know how to add in the proposed changes. Could someone give me some pointers please?

**Comment by joepavitt** [comment]
The PR #1123 is available on the branch `1122-dynamic-props-rewrite` - so you can switch to that branch in your local development environment to get the changes.

**Comment by arturv2000** [comment]
A step by step (at least what I do)

Assuming the git repository is not yet local...

- Create a folder in disk, open a command line and navigate to that folder
- Execute
```
git clone https://github.com/FlowFuse/node-red-dashboard.git
```
will create a folder `node-red-dashboard` or
```
git clone https://github.com/FlowFuse/node-red-dashboard.git .
```
to clone the repo directly to the current folder

- change to the downloaded folder `cd node-red-dashboard`
- Execute `npm install`
- Execute ` git checkout 1122-dynamic-props-rewrite` for this particular branch
![image](https://github.com/user-attachments/assets/a55684c4-bd84-4339-838c-1b2190829c43)
Highlight in red the branch name for this particular PR.
- It can be a good ideia to run `npm install` after checking out a branch, in case it updates dependencies in `package.json`, but typically not necessary.
- Execute `npm run build` or `npm run build:dev` to easally debug in the browser
    - In Windows for dev build first execute `$env:NODE_ENV="development"` and build using `npm run build`
- Open a command line, and navigate to your `node-red` install directory
- Remove the "official" dashboard, `npm uninstall @flowfuse/node-red-dashboard`
- Install the locally compiled version, `npm install D:\Projectos\Node-Red\Main_Repo\` The path should the path where the repository was cloned

After this is necessary to restart `node-red`.

The last two steps are not necessary to repeat if recompiling the project after some modifications, restarting `node-red` may be necessary if changing stuff in the `nodes` folder, and is necessary to at least refresh the dashboard page if recompiling.

**Comment by colinl** [comment]
Thanks @arturv2000. Though I knew most of that. I have been running with a fork, which doesn't have the branch.  In fact rather than work out how to get the branch I cloned from flowfuse directly as you suggest.

@joepavitt the only possibly breaking change I can see is that widget-load now sends `null` the first time, which could cause issues, though it is not a problem for my gauge node, which is the only one on public release.
I am seeing a message in the browser console, though, that I don't remember seeing before
`[Intervention] Slow network is detected. See ￼https://www.chromestatus.com/feature/5636954674692096 for more details. Fallback font will be used while loading: ￼http://localhost:1880/dashboard/assets/materialdesignicons-webfont-Dp5v-WZN.woff2?v=7.4.47`

**Comment by arturv2000** [comment]
@colinl I suffer the same issue, i have searched how to pull a PR from the fork but the instructions seem confusing, so for the safe side i prefer to maintain two repos on my PC, a "myFork" and "Original_Repo" and switch between booth of them as necessary.

**Comment by arturv2000** [comment]
Did some testing, all seems good. At least the uses cases that I was complaining seem to have been solved by this.

**Comment by bartbutenaers** [comment]
Hi @joepavitt,
I have a (yearly reoccuring) computerless period, so I cannot install or test any of this. But it looks quite a great step forward.  So thanks for all your efforts and hard labour!!!

Forgive me if me if my below assumptions are completely incorrect, because reviewing this PR is FAR from easy on a small phone screen 😞

+ Am I correct that we need to update the server and client side state stores ourselves (via quite similar duplicate code), or did you perhaps found a solution to that?  Reason of my question is that in my svg node there are a lot of possible commands, which would result in lots of duplicate code.  And I cannot simply sync the entire state, because that would ear a lot of bandwith.  I did not find a solution for that yet...

+ It "looks" to me that things like ´options´ or ´selection´ (e.g. for a dropdown) are not stored in the state store, but instead are still being stored in the (last msg inside the) data store.  If correct, is there any advantage in not putting everything inside the state store? 

Thanks again for taking into account our feedback!  Really appreciated!
Bart

**Comment by joepavitt** [comment]
> Am I correct that we need to update the server and client side state stores ourselves (via quite similar duplicate code), or did you perhaps found a solution to that? Reason of my question is that in my svg node there are a lot of possible commands, which would result in lots of duplicate code. And I cannot simply sync the entire state, because that would ear a lot of bandwith. I did not find a solution for that yet...

You'll have access to the data tracker now via https://github.com/FlowFuse/node-red-dashboard/pull/1097, so if you're doing the "standard" store of latest message, then no need. Any loading of state will also be handled automatically, and now available in `widget-load` event.

The re-loading of `msg` history, or latest `msg` still stands (currently). I think we'll likely update some APIs there so that it can behave more like the state store, but haven't established any standard there yet.

> It "looks" to me that things like ´options´ or ´selection´ (e.g. for a dropdown) are not stored in the state store, but instead are still being stored in the (last msg inside the) data store. If correct, is there any advantage in not putting everything inside the state store?

`options` still are in the state store. It'll be a recommended pattern that _any_ overrides of the Node-RED configuration should be saved in the `state` store. Only `msg.payload` (value) type data is dependent upon being in the data store.

**Comment by colinl** [comment]
Did you remember that you were going to add some stuff to the example node to show the techniques you are suggesting?

**Comment by joepavitt** [comment]
> Did you remember that you were going to add some stuff to the example node to show the techniques you are suggesting?

Yes Colin 👍

**Comment by colinl** [comment]
I didn't really think you had forgotten :)

**Comment by bartbutenaers** [comment]
> Only msg.payload (value) type data is dependent upon being in the data store

So the ´msg.payload´ is used for real data (e.g. chart samples, radiobutton selected value, ...).  And other state related stuff needs to be injected via ´msg.ui_update´.  If so, it seems to make sense...
